### PR TITLE
PR: Enable LSP folding support

### LINKED ISSUE #9956: Code folding bug
## Problem
When a line of code in a function shows a line break, at this point, the entire function is folded and a bug occurs.
code like this:
115D652E-801A-4ae5-93EE-B33A94DED625
The last tow lines are not folded:
image
But when the line break is deleted, everything is ok.Code like this:
image

Everything is ok:
image
## Versions

* Spyder version: 4.0.0b4
* Python version: 3.7
* Qt version: 
* PyQt version: 
* Operating System name/version: win7

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
This PR enables Spyder to request code folding ranges (if available) for any LSP server, it also removes our old, regex-based code folding.

It depends on
- https://github.com/palantir/python-language-server/pull/665 
- https://github.com/palantir/python-language-server/pull/692 

* [ ] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #9956.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
@andfoy

### CODE REVIEW

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 1]
```diff
@@ -1231,6 +1231,17 @@ def handle_go_to_definition(self, position):
...
+    # ------------- LSP: Code folding ranges -------------------------------
+    @request(method=LSPRequestTypes.DOCUMENT_FOLDING_RANGE)
+    def request_folding(self):
+        params = {'file': self.filename}
+        return params
+
+    @handles(LSPRequestTypes.DOCUMENT_FOLDING_RANGE)
+    def handle_folding_range(self, response):
+        ranges = response['params']
+        logger.debug(f'Ranges: {ranges}')
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/codeeditor.py]: f-strings are only available in Python 3.6+. Please use a formatting method available in Python 2.7 and 3.5+.

**Comment by andfoy** [comment]
ezgif com-video-to-gif(14)

@ccordoba12 this PR is only missing handling editions and folding deletions, but overall, it works as expected

**Comment by andfoy** [comment]
I've just added Intervaltree as an additional dependency, it is available at conda-forge: https://anaconda.org/conda-forge/intervaltree

**Comment by ccordoba12** [comment]
> I've just added Intervaltree as an additional dependency

I added it to our channel, but tests are still failing. Please take a look at them.

**Comment by andfoy** [comment]
@ccordoba12 I've just asked intervaltree maintainers to declare the package version on their ``__init__.py``, otherwise the package versioning checker would always complain, unless we unpin the minimum version required

**Comment by andfoy** [comment]
We depend on https://github.com/palantir/python-language-server/pull/692 for tests to pass

**Comment by ccordoba12** [comment]
@andfoy, I helped you to make the folding tests pass on Windows. I also tested this PR locally and folding is working exceptionally well. Thanks a lot for that!!

However, there are still several important issues to solve before merging:

1. The changes you made to the folding panel broke the debugger panel. What I can see is that the mapping between lines in the editor and those in the debugger panel are affected by the presence of folding arrows. In the screenshot below I'm trying to add a breakpoint on line 20 with the cursor, but it gets added to line 22 after clicking on the debugger panel:

    Selección_003

2. The first line of the folding region is not included in the selected region to be folded. In the screenshot below you can see that I'm trying to fold the block `if running_under_pytest():`, but that line is not included in the selection: 

    Selección_004

    That looks kind of odd from the user perspective, so we need to fix it.

3. Saving a file in this branch (e.g. `spyder/app/mainwindow.py`) generates tons of `textDocument/foldingRange`, which makes saving files a very expensive process. I'm attaching a screenshot below of what I can see in the console after starting Spyder with `python bootstrap.py --debug`:

    Selección_005

**Comment by andfoy** [comment]
@ccordoba12 Does the last commit fixes the aforementioned issues?

**Comment by ccordoba12** [comment]
@andfoy, the three issues I mentioned above are solved now. Thanks a lot for that!

However, I found some issues with our indentation guides support, which you also rewrote as part of this PR. Below you can find a couple of examples in `mainwindow.py` where those lines are drawn on top of text, which looks very ugly.

Selección_007

Selección_006

Since this is not happening in master, it's a regression introduced by your work. Please see how to fix it.

**Comment by andfoy** [comment]
> However, I found some issues with our indentation guides support, which you also rewrote as part of this PR. Below you can find a couple of examples in mainwindow.py where those lines are drawn on top of text, which looks very ugly.

Yes indeed, this is due to the non-constant indentation used by some files.

**Comment by ccordoba12** [comment]
But do you think you can fix it?

**Comment by andfoy** [comment]
> But do you think you can fix it?

We need to compute the actual line space offset before drawing the lines, that is easy when using spaces, but not when tabs and spaces are mixed. I'll try to come up with a solution

**Comment by Khalilsqu** [comment]
I think this PR might close this [issue](https://github.com/spyder-ide/spyder/issues/8864) as well. This PR draws the indentation guides as one line for each fold, while the original one drew the lines for each line within the fold level (computationally expensive)

**Comment by ccordoba12** [comment]
@CAM-Gerlach, could you test this PR to see if it fixes your problem with indentation guides?

**Comment by Khalilsqu** [comment]
@andfoy there is an empty line at the start of the file `indentationguides.py` which was introduced by me mistakenly. Would you mind to remove it if possible as part of this PR!

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Sorry for the delay; in work now.

File: requirements/conda.txt [hunk 2]
```diff
@@ -25,3 +25,4 @@ rtree
 sphinx
 spyder-kernels >=1.7.0
 watchdog
+intervaltree
```
Comment (CAM-Gerlach) [inline requirements/conda.txt]: Not alphabetized.

Comment (ccordoba12) [reply requirements/conda.txt]: I agree. @andfoy, please put this new dependency in alphabetical order among the other ones.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 @andfoy I installed the `intervaltree` dependency (and the others changed over rc1, as well as ensuring I had the most recently released versions of Spyder-Kernels, LSP, qt/pyqt/qtpy, python, and  the other main deps) and jedi 0.14.1). However, when enabling code folding when running from this branch, even if I close and reopen files/projects/etc and check small or large files, no indent guides appear and furthermore code folding no longer works, the arrows and indications simply don't appear at all.

palantir/python-language-server#665 seems to suggest that pyls 0.30 is needed to enable this feature, but 0.29.3, which I have, is the latest release available on the `spyder-ide` channel and furthermore both the requirements file and setup.py in this PR still explicitly prohibit pyls <= 0.30 (like `master`). @ccordoba12 ?

**Comment by ccordoba12** [comment]
Thanks your time @CAM-Gerlach. I'll let you when the new dependencies required by this PR are available in our channel for you test it.

**Comment by CAM-Gerlach** [comment]
Sure, thanks @ccordoba12 . Just let me know when you'd like me to test again.

**Comment by andfoy** [comment]
@ccordoba12 This one is ready for a final review

**Comment by CAM-Gerlach** [comment]
Testing in work.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Let me know when you've got the LSP 0.30+ packages up on `spyder-ide` if you'd like me to test to see if this fixes #8864 and works well on Windows.

@andfoy As I mentioned previously, don't you need to update the LSP version spec in the requirements files and the setup.py so this doesn't just break code folding and indentation in Spyder? Right now, both of those block the appropriate version of LSP (0.30) from being installed, instead of requiring it.

**Comment by ccordoba12** [comment]
@andfoy, things are looking much better now, thanks!

However, there are still errors. These are from `mainwindown.py` again, using the latest commit in this branch:

Selección_020

Selección_022

Selección_021

The first and second screenshots are very similar, in the sense that spurious lines are introduced due to strange indentations. The last one is showing a line on top of a docstring, which I think looks really bad.

**Comment by andfoy** [comment]
The aforementioned case about overpainting is due to an error on the folding computing, rather than by the painting itself, this issue will be fixed on the pyls branch this PR depends on

**Comment by andfoy** [comment]
We could remove the line for docstrings

**Comment by ccordoba12** [comment]
> The aforementioned case about overpainting is due to an error on the folding computing, rather than by the painting itself, this issue will be fixed on the pyls branch this PR depends on

Ok, thanks a lot for that!

> We could remove the line for docstrings

Ok, I agree with that.

**Comment by CAM-Gerlach** [comment]
> We could remove the line for docstrings

And by the same token, it should be not be shown for any multi-line `"""` or `'''` string literal, for conceptual as well as the aforementioned aesthetic reasons. Since docstrings, or really, multiline strings in genera, do not explicitly define another indentation level or code scope (they are just a single long literal). then I'm not sure the line should be there conceptually at all. In my limited knowledge, I'm not aware of another popular IDE/editor that implements indentation guides this way (even though they usually do allow code folding a docstring).

**Comment by andfoy** [comment]
@ccordoba12 Does the last commits on the pyls PR fix the latest issues?

**Comment by ccordoba12** [comment]
> Does the last commits on the pyls PR fix the latest issues?

I don't see indentation lines in docstrings now, but I'm still seeing spurious lines in the two screenshots I posted in my https://github.com/spyder-ide/spyder/pull/10333#issuecomment-554798140 above.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, this is ready for you to review it. Please install the new python-language-server package available in our channel and give it a try. It doesn't contain the latest fixes by @andfoy to deal with spurious lines, but it should solve your problem with performance issues.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, I'm going to merge this one because it's a great improvement. So please comment on issue #8864 if it fixes that or not.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Sorry, I try to give reviews my highest priority but a 7 minute latency is a bit much even for me :) I'll test it and comment on #8864 .

**Pull request merged**